## 1. Import Libraries


In [1]:
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

## 2. Reading data and dropping unwanted columns

In [2]:
PROJECT_DIRECTORY = r"C:\Users\msved\OneDrive\Documents\Campus X\FLIGHT PREDICTION USING AWS SAGEMAKER"
DATA_DIRECTORY = "data"
# here, r is for raw string
''' In Python, when you see a string prefixed with r, such as r"some_string", it denotes a raw string literal.
A raw string literal treats backslashes (\) as literal characters and not as escape characters.'''

' In Python, when you see a string prefixed with r, such as r"some_string", it denotes a raw string literal.\nA raw string literal treats backslashes (\\) as literal characters and not as escape characters.'

In [3]:
!pip install pandas openpyxl
# this is for to run fetch_data function

In [4]:
def fetch_data(name):
    file_name = f"{name}.xlsx"
    file_path = os.path.join(PROJECT_DIRECTORY,DATA_DIRECTORY,file_name)
    return pd.read_excel(file_path)

In [5]:
flight_data = fetch_data("Flight_Fare_Prediction")
flight_data

Date_of_Journey   Airlines Airlines_code     Class Departure  \
0          2024-02-29     Indigo       6E-2519   Economy     09:03   
1          2024-06-25   GO FIRST        G8-354   Economy     18:25   
2          2024-04-07   SpiceJet       SG-8709   Economy     12:14   
3          2024-06-20  Air India        AI-805   Economy     02:37   
4          2024-04-07  Air India        AI-605   Economy     17:18   
...               ...        ...           ...       ...       ...   
10677      2024-07-01  Air India        AI-811  Business     00:21   
10678      2024-05-17  Air India        AI-479  Business     01:27   
10679      2024-08-22  Air India        AI-463  Business     06:10   
10680      2024-10-31  Air India        AI-471  Business     14:36   
10681      2024-08-08    Vistara        UK-635  Business     02:30   

      Total_Stops Arrival  Duration_in_Hours    Source Destination  \
0        non-stop   05:03             2.3333  Banglore   New Delhi   
1         2 stops   20:26             2.1667   Kolkata    Banglore   
2         2 stops   21:25             2.0833     Delhi      Cochin   
3          1 stop   21:44             2.1667   Kolkata    Banglore   
4          1 stop   00:08             2.2500  Banglore   New Delhi   
...           ...     ...                ...       ...         ...   
10677    non-stop   12:16            21.7500  Banglore       Delhi   
10678    non-stop   00:19            20.6667   Kolkata    Banglore   
10679    non-stop   18:12            27.6667   Kolkata    Banglore   
10680    non-stop   09:50             5.3333  Banglore       Delhi   
10681    non-stop   09:55             4.8333  Banglore   New Delhi   

                       Route  Price_fare Unnamed: 12  
0                  BLR → DEL        5899  2024-01-01  
1      CCU → IXR → BBI → BLR        5801  2024-12-31  
2      DEL → LKO → BOM → COK        5794         NaT  
3            CCU → NAG → BLR        5955         NaT  
4            BLR → NAG → DEL        5955         NaT  
...                      ...         ...         ...  
10677              BLR → DEL       32646         NaT  
10678              CCU → BLR       35532         NaT  
10679              CCU → BLR       35532         NaT  
10680              BLR → DEL       37256         NaT  
10681              BLR → DEL       37417         NaT  

[10682 rows x 13 columns]

In [6]:
# removing unwanted columns
if 'Unnamed: 12' in flight_data.columns:
    flight_data.drop('Unnamed: 12', axis=1, inplace=True)
if 'Unnamed: 13' in flight_data.columns:
    flight_data.drop('Unnamed: 13', axis=1, inplace=True)

In [7]:
# after removing 2 columns, our data looks like this.
flight_data

Date_of_Journey   Airlines Airlines_code     Class Departure  \
0          2024-02-29     Indigo       6E-2519   Economy     09:03   
1          2024-06-25   GO FIRST        G8-354   Economy     18:25   
2          2024-04-07   SpiceJet       SG-8709   Economy     12:14   
3          2024-06-20  Air India        AI-805   Economy     02:37   
4          2024-04-07  Air India        AI-605   Economy     17:18   
...               ...        ...           ...       ...       ...   
10677      2024-07-01  Air India        AI-811  Business     00:21   
10678      2024-05-17  Air India        AI-479  Business     01:27   
10679      2024-08-22  Air India        AI-463  Business     06:10   
10680      2024-10-31  Air India        AI-471  Business     14:36   
10681      2024-08-08    Vistara        UK-635  Business     02:30   

      Total_Stops Arrival  Duration_in_Hours    Source Destination  \
0        non-stop   05:03             2.3333  Banglore   New Delhi   
1         2 stops   20:26             2.1667   Kolkata    Banglore   
2         2 stops   21:25             2.0833     Delhi      Cochin   
3          1 stop   21:44             2.1667   Kolkata    Banglore   
4          1 stop   00:08             2.2500  Banglore   New Delhi   
...           ...     ...                ...       ...         ...   
10677    non-stop   12:16            21.7500  Banglore       Delhi   
10678    non-stop   00:19            20.6667   Kolkata    Banglore   
10679    non-stop   18:12            27.6667   Kolkata    Banglore   
10680    non-stop   09:50             5.3333  Banglore       Delhi   
10681    non-stop   09:55             4.8333  Banglore   New Delhi   

                       Route  Price_fare  
0                  BLR → DEL        5899  
1      CCU → IXR → BBI → BLR        5801  
2      DEL → LKO → BOM → COK        5794  
3            CCU → NAG → BLR        5955  
4            BLR → NAG → DEL        5955  
...                      ...         ...  
10677              BLR → DEL       32646  
10678              CCU → BLR       35532  
10679              CCU → BLR       35532  
10680              BLR → DEL       37256  
10681              BLR → DEL       37417  

[10682 rows x 12 columns]

In [8]:
flight_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10682 entries, 0 to 10681
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Date_of_Journey    10682 non-null  datetime64[ns]
 1   Airlines           10682 non-null  object        
 2   Airlines_code      10682 non-null  object        
 3   Class              10682 non-null  object        
 4   Departure          10682 non-null  object        
 5   Total_Stops        10681 non-null  object        
 6   Arrival            10682 non-null  object        
 7   Duration_in_Hours  10682 non-null  float64       
 8   Source             10682 non-null  object        
 9   Destination        10682 non-null  object        
 10  Route              10681 non-null  object        
 11  Price_fare         10682 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(1), object(9)
memory usage: 1001.6+ KB


- Here, we have 1 missing values in Total_Stops and Route column.
- We have total 10682 rows and 12 columns


# 3 Preliminary Analysis


## 3.1 checking data types

In [9]:
flight_data.dtypes

Date_of_Journey      datetime64[ns]
Airlines                     object
Airlines_code                object
Class                        object
Departure                    object
Total_Stops                  object
Arrival                      object
Duration_in_Hours           float64
Source                       object
Destination                  object
Route                        object
Price_fare                    int64
dtype: object

In [10]:
value = flight_data.Date_of_Journey.iloc[19]
value


Timestamp('2024-10-21 00:00:00')

In [11]:
value1 = flight_data.Departure.iloc[5]
value1

'09:04'

In [12]:
value2 = flight_data.Arrival.iloc[10]
value2

'19:48'

In [13]:
value3 = flight_data.Duration_in_Hours.iloc[10]
value3

8.6667

In [14]:
value4 = flight_data.Total_Stops.iloc[10]
value4

'1 stop'

## 3.2 Check for duplicates

In [15]:
flight_data.duplicated().sum()

0

In [16]:
(
flight_data
    .loc[flight_data.duplicated(keep=False)]
    .sort_values(['Airlines','Date_of_Journey','Source','Destination'])
)
# loc function is used to filter an object in python.
# Keep=False means it also counts the first occurance of the datasets.

Empty DataFrame
Columns: [Date_of_Journey, Airlines, Airlines_code, Class, Departure, Total_Stops, Arrival, Duration_in_Hours, Source, Destination, Route, Price_fare]
Index: []

## 3.3 Observations

- 1.Departure, and 2.Arrival time should be data type of date not string.
- 3. Total_Stops should be numeric data type not in the form of string
- 4. There are 9 duplicates rows and it should be remove.

# 4. Analysis in Detail

In [17]:
flight_data.Airlines

0           Indigo
1         GO FIRST
2         SpiceJet
3        Air India
4        Air India
           ...    
10677    Air India
10678    Air India
10679    Air India
10680    Air India
10681      Vistara
Name: Airlines, Length: 10682, dtype: object

In [18]:
flight_data.Airlines.unique()

array(['Indigo', 'GO FIRST', 'SpiceJet', 'Air India', 'AirAsia',
       'Vistara', 'AkasaAir'], dtype=object)

In [19]:

(
flight_data
    .Airlines
    .str.replace("FIRST","First")
    .str.replace("AirAsia","Air asia")
    .str.replace("AkasaAir","Akasa Air")
    .str.title()
    .unique()
)

array(['Indigo', 'Go First', 'Spicejet', 'Air India', 'Air Asia',
       'Vistara', 'Akasa Air'], dtype=object)

- There is no inconsistent result in Airlines column

In [20]:
flight_data

Date_of_Journey   Airlines Airlines_code     Class Departure  \
0          2024-02-29     Indigo       6E-2519   Economy     09:03   
1          2024-06-25   GO FIRST        G8-354   Economy     18:25   
2          2024-04-07   SpiceJet       SG-8709   Economy     12:14   
3          2024-06-20  Air India        AI-805   Economy     02:37   
4          2024-04-07  Air India        AI-605   Economy     17:18   
...               ...        ...           ...       ...       ...   
10677      2024-07-01  Air India        AI-811  Business     00:21   
10678      2024-05-17  Air India        AI-479  Business     01:27   
10679      2024-08-22  Air India        AI-463  Business     06:10   
10680      2024-10-31  Air India        AI-471  Business     14:36   
10681      2024-08-08    Vistara        UK-635  Business     02:30   

      Total_Stops Arrival  Duration_in_Hours    Source Destination  \
0        non-stop   05:03             2.3333  Banglore   New Delhi   
1         2 stops   20:26             2.1667   Kolkata    Banglore   
2         2 stops   21:25             2.0833     Delhi      Cochin   
3          1 stop   21:44             2.1667   Kolkata    Banglore   
4          1 stop   00:08             2.2500  Banglore   New Delhi   
...           ...     ...                ...       ...         ...   
10677    non-stop   12:16            21.7500  Banglore       Delhi   
10678    non-stop   00:19            20.6667   Kolkata    Banglore   
10679    non-stop   18:12            27.6667   Kolkata    Banglore   
10680    non-stop   09:50             5.3333  Banglore       Delhi   
10681    non-stop   09:55             4.8333  Banglore   New Delhi   

                       Route  Price_fare  
0                  BLR → DEL        5899  
1      CCU → IXR → BBI → BLR        5801  
2      DEL → LKO → BOM → COK        5794  
3            CCU → NAG → BLR        5955  
4            BLR → NAG → DEL        5955  
...                      ...         ...  
10677              BLR → DEL       32646  
10678              CCU → BLR       35532  
10679              CCU → BLR       35532  
10680              BLR → DEL       37256  
10681              BLR → DEL       37417  

[10682 rows x 12 columns]

### Source

In [21]:
flight_data.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [22]:
flight_data.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [23]:
# checking if syntax of departure time is wrong or not
(
flight_data
    .Departure
    .loc[lambda ser : ser.str.contains("[^0-9:]")]
)

Series([], Name: Departure, dtype: object)

In [24]:
# checking if syntax of arrival time is wrong or not
(
flight_data
    .Arrival
    .loc[lambda ser : ser.str.contains("[^0-9:]")]
)

Series([], Name: Arrival, dtype: object)

In [25]:
(
flight_data
    .Total_Stops
    .str.replace("non-stop","0")
    .str.replace("stops?","",regex=True) #here, last s is optional otherwise, it always check for stop
    #.str.replace("stop","")
    .pipe(lambda ser : pd.to_numeric(ser))
    .unique()
)

array([ 0.,  2.,  1.,  3., nan,  4.])

# 5. Cleaning Operations

In [26]:
def data_clean(df):
    return (
        
    df
    .drop_duplicates()
    .assign(**{
        col : df[col].str.strip() 
        for col in df.select_dtypes(include='O').columns
    })
    .rename(columns = str.lower)
    .assign(airlines = lambda df_: (
        df_.airlines
        .str.replace("FIRST","First")
        .str.replace("AirAsia","Air asia")
        .str.replace("AkasaAir","Akasa Air")
        .str.title()
    ),
            departure = lambda df_ : pd.to_datetime(df_.departure).dt.time,
            arrival = lambda df_ : pd.to_datetime(df_.arrival).dt.time,
            duration_in_hours = lambda df_ : df_.duration_in_hours.round(2)*60,
            total_stops = lambda df_ : df_.total_stops.str.replace("non-stop","0").str.replace("stops?","",regex=True)
            #here, last s is optional otherwise, it always check for stop
            #.str.replace("stop","")
            .pipe(lambda ser : pd.to_numeric(ser))
            
           )
        .drop(columns="route")
        .rename(columns={'duration_in_hours': 'duration_in_minutes'})
        .fillna(method='ffill')
    )
# ** is used for unpacking
# also, dictionary is written in assign function and "O" means object data type
''' dt.time is used to get time only, otherwise to_datetime function will show date and time both,
 however, we don't require date, therefore we used dt.time. '''
# lambda is used to access current data frame
# he .fillna(method='ffill') method call with the argument method='ffill' is used to fill missing values in a DataFrame using forward fill method. Here's what it does:

# Forward Fill (ffill): This method fills missing values with the last known value along the specified axis. 
# For each missing value, it takes the value from the previous non-missing (valid) value in the same column.

" dt.time is used to get time only, otherwise to_datetime function will show date and time both,\n however, we don't require date, therefore we used dt.time. "

In [27]:
data_clean(flight_data).airlines.unique()

array(['Indigo', 'Go First', 'Spicejet', 'Air India', 'Air Asia',
       'Vistara', 'Akasa Air'], dtype=object)

In [28]:
data_clean(flight_data).duration_in_minutes.min()
# min time duration in minutes

115.19999999999999

In [29]:
flight_cleaned_data = data_clean(flight_data)
flight_cleaned_data

date_of_journey   airlines airlines_code     class departure  \
0          2024-02-29     Indigo       6E-2519   Economy  09:03:00   
1          2024-06-25   Go First        G8-354   Economy  18:25:00   
2          2024-04-07   Spicejet       SG-8709   Economy  12:14:00   
3          2024-06-20  Air India        AI-805   Economy  02:37:00   
4          2024-04-07  Air India        AI-605   Economy  17:18:00   
...               ...        ...           ...       ...       ...   
10677      2024-07-01  Air India        AI-811  Business  00:21:00   
10678      2024-05-17  Air India        AI-479  Business  01:27:00   
10679      2024-08-22  Air India        AI-463  Business  06:10:00   
10680      2024-10-31  Air India        AI-471  Business  14:36:00   
10681      2024-08-08    Vistara        UK-635  Business  02:30:00   

       total_stops   arrival  duration_in_minutes    source destination  \
0              0.0  05:03:00                139.8  Banglore   New Delhi   
1              2.0  20:26:00                130.2   Kolkata    Banglore   
2              2.0  21:25:00                124.8     Delhi      Cochin   
3              1.0  21:44:00                130.2   Kolkata    Banglore   
4              1.0  00:08:00                135.0  Banglore   New Delhi   
...            ...       ...                  ...       ...         ...   
10677          0.0  12:16:00               1305.0  Banglore       Delhi   
10678          0.0  00:19:00               1240.2   Kolkata    Banglore   
10679          0.0  18:12:00               1660.2   Kolkata    Banglore   
10680          0.0  09:50:00                319.8  Banglore       Delhi   
10681          0.0  09:55:00                289.8  Banglore   New Delhi   

       price_fare  
0            5899  
1            5801  
2            5794  
3            5955  
4            5955  
...           ...  
10677       32646  
10678       35532  
10679       35532  
10680       37256  
10681       37417  

[10682 rows x 11 columns]

In [30]:
flight_cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10682 entries, 0 to 10681
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   date_of_journey      10682 non-null  datetime64[ns]
 1   airlines             10682 non-null  object        
 2   airlines_code        10682 non-null  object        
 3   class                10682 non-null  object        
 4   departure            10682 non-null  object        
 5   total_stops          10682 non-null  float64       
 6   arrival              10682 non-null  object        
 7   duration_in_minutes  10682 non-null  float64       
 8   source               10682 non-null  object        
 9   destination          10682 non-null  object        
 10  price_fare           10682 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(1), object(7)
memory usage: 1001.4+ KB


# 6. Split Database into train data and test data

In [31]:
X = flight_cleaned_data.drop(columns = "price_fare")
y = flight_cleaned_data.price_fare.copy() 

In [32]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(6836, 10) (6836,)
(1709, 10) (1709,)
(2137, 10) (2137,)


# 7. Subset

In [33]:
def data_export(X,y,name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIRECTORY, DATA_DIRECTORY,file_name)
    X.join(y).to_csv(file_path,index=False)
    return pd.read_csv(file_path).head()

In [34]:
data_export(X_train, y_train , "train")

date_of_journey   airlines airlines_code            class departure  \
0      2024-04-01    Vistara        UK-855  Premium Economy  00:26:00   
1      2024-03-26  Air India        AI-443          Economy  12:19:00   
2      2024-12-09  Air India        AI-815         Business  10:46:00   
3      2024-09-08   Go First       G8-2519          Economy  21:06:00   
4      2024-05-04    Vistara        UK-929         Business  18:15:00   

   total_stops   arrival  duration_in_minutes    source destination  \
0          1.0  07:27:00                334.8     Delhi      Cochin   
1          1.0  05:17:00                304.8   Kolkata    Banglore   
2          1.0  18:57:00                355.2     Delhi      Cochin   
3          0.0  13:02:00                850.2  Banglore       Delhi   
4          1.0  02:34:00                535.2   Kolkata    Banglore   

   price_fare  
0       23238  
1        8516  
2       53164  
3        6873  
4       22297

In [35]:
data_export(X_val, y_val , "validation")

date_of_journey   airlines airlines_code     class departure  total_stops  \
0      2024-02-11  Air India        AI-665  Business  12:30:00          1.0   
1      2024-01-07  Akasa Air       QP-1128   Economy  19:28:00          1.0   
2      2024-06-22    Vistara        UK-979  Business  19:11:00          1.0   
3      2024-11-13  Air India        AI-624  Business  04:35:00          1.0   
4      2024-02-09     Indigo       6E-6492   Economy  19:41:00          0.0   

    arrival  duration_in_minutes    source destination  price_fare  
0  21:58:00                139.8   Kolkata    Banglore       22152  
1  12:10:00                130.2   Kolkata    Banglore        5890  
2  19:56:00               1009.8     Delhi      Cochin       30249  
3  19:48:00                145.2     Delhi      Cochin       22152  
4  02:03:00                319.8  Banglore       Delhi        5962

In [36]:
data_export(X_test, y_test , "test")

date_of_journey airlines airlines_code            class departure  \
0      2024-11-27   Indigo       6E-6814          Economy  05:04:00   
1      2024-05-27  Vistara        UK-785  Premium Economy  05:05:00   
2      2024-11-17  Vistara        UK-817          Economy  12:18:00   
3      2024-07-23  Vistara        UK-891          Economy  17:46:00   
4      2024-04-01  Vistara        UK-879         Business  22:39:00   

   total_stops   arrival  duration_in_minutes   source destination  price_fare  
0          2.0  22:07:00                145.2    Delhi      Cochin        5954  
1          1.0  08:46:00               1065.0    Delhi      Cochin       18534  
2          0.0  06:50:00                435.0   Mumbai   Hyderabad        7189  
3          0.0  01:40:00                454.8  Chennai     Kolkata       20340  
4          1.0  19:55:00                880.2    Delhi      Cochin       55696